test/common.py

```diff
@@ -375,6 +347,61 @@ def run_train_translate(train_params: str,
         with patch.object(sys, "argv", params.split()):
             sockeye.translate.main()
 
+        # Test target constraints
+        if use_target_constraints:
+            """
+            Read in the unconstrained system output from the first pass and use it to generate positive
+            and negative constraints. It is important to generate a mix of positive, negative, and no
+            constraints per batch, to test these production-realistic interactions as well.
+            """
+            # 'constraint' = positive constraints (must appear), 'avoid' = negative constraints (must not appear)
+            for constraint_type in ["constraints", "avoid"]:
+                constrained_sources = []
+                with open(test_source_path) as source_inp, open(out_path) as system_out:
+                    for sentno, (source, target) in enumerate(zip(source_inp, system_out)):
+                        target_words = target.rstrip().split()
+                        target_len = len(target_words)
+                        new_source = {'text': source.rstrip()}
+                        # From the odd-numbered sentences that are not too long, create constraints. We do
+                        # only odds to ensure we get batches with mixed constraints / lack of constraints.
+                        if target_len > 0 and sentno % 2 == 0:
+                            start_pos = 0
+                            end_pos = min(target_len, 3)
+                            constraint = ' '.join(target_words[start_pos:end_pos])
+                            new_source[constraint_type] = [constraint]
+                        constrained_sources.append(json.dumps(new_source))
+
+                new_test_source_path = os.path.join(work_dir, "test_constrained.txt")
+                with open(new_test_source_path, 'w') as out:
+                    for json_line in constrained_sources:
+                        print(json_line, file=out)
+
+                out_path_constrained = os.path.join(work_dir, "out_constrained.txt")
+                params = "{} {} {} --json-input".format(sockeye.translate.__file__,
+                                                        _TRANSLATE_PARAMS_COMMON.format(model=model_path,
+                                                                                        input=new_test_source_path,
+                                                                                        output=out_path_constrained,
+                                                                                        quiet=quiet_arg),
+                                                        translate_params)
+
+                with patch.object(sys, "argv", params.split()):
+                    sockeye.translate.main()
+
+                print('JSON_INPUT', open(new_test_source_path).readlines(),
```

Comment:
could we change this to a logging statement? with pytest, logging statements are easier to track than stdout output.

Reply:
Oh, this should be removed entirely.